enhancements/20240528-migrate-harvester-api-to-standardized-k8s-api.md

```diff
@@ -0,0 +1,159 @@
+# Migrate harvester API actions to standardized k8s API
+
+
+## Summary
+
+Harvester currently use rancher/steve to build custom api actions, such as stop vm with `?action=stop` which is not standard k8s API. 
+
+We will leverage the Kubernetes API Aggregation Layer, which proxies client requests to a specified API server defined with a custom resource `APIService`.
+
+This allows us to use the native Kubernetes client to invoke subresource APIs. 
+
+```go
+harvesterSubresourceClient.
+	Post().
+	Namespace("default").
+	Resource("virtualmachines").
+	SubResource("stop").
+	Name("test").Do(context.Background())
+```
+
+### Related Issue
+
+- https://github.com/harvester/harvester/issues/1078
+  - This is root issue.
+- https://github.com/harvester/harvester/issues/5627
+  - This is sub-issue about we need to add documentation for the action APIs. 
+
+
+## Motivation
+
+### Goals
+
+- Ensure the API format aligns with Kubernetes conventions for better consistency and security.
+- Support old and new API actions at the same time for backward compatibility.
+
+### Non-Goals
+
+- Remove old API actions immediately.
+- Migrate current non action APIs to k8s standardized APIs.
+
+## Proposal
+
+### Use Case 1 - API Usage
+
+In real usecase, the old api looks like this:
+
+```
+POST /v1/harvester/kubevirt.io.virtualmachines/default/test?action=stop
+POST /v1/harvester/nodes/jacknode?action=cordon
+GET  /v1/harvester/upgradeslogs/default/test/download
+POST /v1/harvester/persistentvolumeclaims/default/test-disk-0-rlnlk?action=snapshot
+```
+
+The new api will be like this:
+
+```yaml
+POST /apis/subresources.harvesterhci.io/v1beta1/namespaces/default/virtualmachines/test/stop
```

Comment:
In places where Harvester utilizes CRDs from other projects, e.g. KubeVirt, there are such API paths already in place: e.g. https://kubevirt.io/api-reference/v1.2.1/operations.html#_v1alpha3stop
We don't need to create duplicate API paths for such actions. It's still a good idea to follow this approach for the Harvester CRDs.

It would be a good thing to compile a complete list of new API routes for this HEP. It would help in estimating the implementation effort and we could double check to make sure we're not duplicating existing API routes.

Reply:
we should definitely start with having a list of apis exposed by harvester and split them into ones which will be moved to harvester subresources and the ones which will be replaced with kubevirt native subresources. We will also need to document the impact of the same on the terraform provider and machine driver if there is any.